File: leetcode/722.cc
#include <string>
#include <vector>


using namespace std;


class Solution {
  public:
    int FindNextOpen(const string& line) {
        auto t1 = line.find("//");
        auto t2 = line.find("/*");
        return min(t1, t2);
    }

    vector<string> removeComments(vector<string>& source) {
        vector<string> result;
        bool block = false;
        for (auto& line : source) {
            bool compact = block;
            int del_start = 0;
            int search_start = 0;
            while (true) {
                if (!block) {
                    auto t = FindNextOpen(line);
                    if (t == string::npos)
                        break;
                    auto token = line.substr(t, 2);
                    if (token == "//") {
                        line.erase(t);
                        break;
                    } else if (token == "/*") {
                        del_start = t;
                        search_start = t + 2;
                        block = true;
                    }
                } else {
                    auto t = line.find("*/", search_start);
                    if (t != string::npos) {
                        line.erase(del_start, t - del_start + 2);
                        block = false;
                    } else {
                        line.erase(del_start);
                        break;
                    }
                }
            }

            if (!line.empty()) {
                if (compact)
                    result.back() += line;
                else
                    result.emplace_back(line);
            }
        }
        return result;
    }
};

File: leetcode/722.py
class Solution:
    def removeComments(self, source: List[str]) -> List[str]:
        block = False
        result = []
        buf = ''

        for line in source:
            i = 0
            t = len(line)
            while i < t:
                if not block:
                    if i < t - 1 and line[i:i+2] == '//':
                        break
                    elif i < t - 1 and line[i:i+2] == '/*':
                        block = True
                        i += 1
                    else:
                        buf += line[i]
                else:
                    if i < t - 1 and line[i:i+2] == '*/':
                        block = False
                        i += 1
                i += 1
            if not block and buf:
                result.append(buf)
                buf = ''
        return result
